WordPress/build.gradle

```diff
@@ -133,7 +135,9 @@ android {
             buildConfigField "String", "PUSH_NOTIFICATIONS_APP_KEY", '"org.wordpress.android"'
 
             manifestPlaceholders = [magicLinkScheme:"wordpress"]
+
         }
+        addBuildConfigFieldsFromPrefixedProperties(wordpress, properties, 'wp')
```

Comment:
My initial goal was to move that call within the `android { productFlavor { wordpress { … } } }` block just above (i.e. move this on line 134/136) and have it implicitly capture the flavor as parameter, but I wasn't able to achieve this. 

I'd love some pointers on how to do that, as the goal here would be to make this as most visible a possible by having that call sit next to the other manual `buildConfigField` declarations, so that it's obvious that other `buildConfigFields` are declared implicitly via properties in addition to the ones done manually, hopefully helping better discover the logic

I tried using `addBuildConfigFieldsFromPrefixedProperties(this, properties, 'wp')` but apparently even in the scope of `android { productFlavors { wordpress { … } } }` block, `this` points to the project, not to the current scope = the workpress flavor. I hope I'm missing something and there's a magix way to capture current scope to pass it to the call so we can move it to the right place?

Reply:
The thing to keep in mind is that, this DSL is not treated as a `switch` code block. The project will be fully evaluated regardless of which task is being run. So, trying to separate the logic per flavor in my opinion makes it harder to understand - especially if the developer didn't know the project will be fully evaluated.

This statement ^ can be confirmed by applying the following diff and running `./gradlew` (which basically just configures the project but doesn't run any tasks):

```
diff --git a/WordPress/build.gradle b/WordPress/build.gradle
index bdae09615a..bd9ca0254d 100644
--- a/WordPress/build.gradle
+++ b/WordPress/build.gradle
@@ -136,6 +136,7 @@ android {
 
             manifestPlaceholders = [magicLinkScheme:"wordpress"]
 
+            println("evaluating WordPress flavor")
         }
         addBuildConfigFieldsFromPrefixedProperties(wordpress, properties, 'wp')
 
@@ -148,6 +149,7 @@ android {
             buildConfigField "String", "PUSH_NOTIFICATIONS_APP_KEY", '"com.jetpack.android"'
 
             manifestPlaceholders = [magicLinkScheme:"jetpack"]
+            println("evaluating jetpack flavor")
 
         }
         addBuildConfigFieldsFromPrefixedProperties(jetpack, properties, 'wp') // Inherit same properties used from WP
@@ -160,24 +162,28 @@ android {
             buildConfigField "boolean", "ENABLE_FEATURE_CONFIGURATION", "false"
             buildConfigField "boolean", "CONTACT_INFO_BLOCK_AVAILABLE", "true"
             buildConfigField "boolean", "LAYOUT_GRID_BLOCK_AVAILABLE", "true"
+            println("evaluating vanilla flavor")
         }
 
         zalpha { // alpha version - enable experimental features
             dimension "buildType"
             buildConfigField "boolean", "VIDEO_OPTIMIZATION_AVAILABLE", "true"
             buildConfigField "boolean", "ENABLE_FEATURE_CONFIGURATION", "false"
+            println("evaluating zalpha flavor")
         }
 
         wasabi { // "hot" version, can be installed along release, alpha or beta versions
             applicationIdSuffix ".beta"
             dimension "buildType"
             buildConfigField "boolean", "UNIFIED_COMMENTS_LIST", "true"
+            println("evaluating wasabi flavor")
         }
 
         jalapeno { // Pre-Alpha version, used for PR builds, can be installed along release, alpha, beta, dev versions
             isDefault true
             applicationIdSuffix ".prealpha"
             dimension "buildType"
+            println("evaluating jalapeno flavor")
         }
     }
```

Reply:
Yeah I know, I just hoped that there was a way to capture the scope when that line was evaluated. Not to capture "which falvor is being build" – since I understand now (thanks to your teachings 🙂 ) that the whole file is evaluated anyway before executing the tasks so that's not how it would work – but rather "during the evaluation of that block, capture whatever context we're in to be able to pass it to the method call".

Basically, when I write `addBuildConfigFieldsFromPrefixedProperties(wordpress, properties, 'wp')` inside the `android { productFlavors { … } }` scope, it is already able to evaluate that the `wordpress` parameter that I reference here is in fact `android.productFlavors.wordpress` because I'm already in the `android.productFlavors` scope so I can just use the relative name rather than having to write `addBuildConfigFieldsFromPrefixedProperties(android.productFlavors.wordpress, properties, 'wp')`.

So my goal was to use that behavior and convenience offered by the DSL (that you can reference properties from the scope like I already did here, rather than their fully-qualified "path"), but on the nested scope.

If I call `addBuildConfigFieldsFromPrefixedProperties(…)` from the root of the DSL, I'll have to use the full `android.productFlavors.wordpress` explicitly; if I call it from within the `android { … }` scope, I can just use `productFlavors.wordpress` to reference the same thing. If I'm within the `android { productFlavors { … } }` like here, I can just reference it as `wordpress` like I did. But if I'm within the `android { productFlavors { wordpress { } } }` scope, what's the keyword I could use to mean `this` then?